From 7fa7976c70fbe949b64602993705e68379cb7f36 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:09:47 -0400
Subject: [virt] kvm/ppc: Clear the runlatch bit of a vcpu before napping

Message-id: <1410545655-205645-359-git-send-email-dzickus@redhat.com>
Patchwork-id: 94267
O-Subject: [RHEL7 PATCH 358/626] ppc/kvm: Clear the runlatch bit of a vcpu before napping
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 582b910edafd283dfab78f41f437a92a65ee5103
Author: Preeti U Murthy <preeti@linux.vnet.ibm.com>
Date:   Fri Apr 11 16:02:08 2014 +0530

    ppc/kvm: Clear the runlatch bit of a vcpu before napping

    When the guest cedes the vcpu or the vcpu has no guest to
    run it naps. Clear the runlatch bit of the vcpu before
    napping to indicate an idle cpu.

    Signed-off-by: Preeti U Murthy <preeti@linux.vnet.ibm.com>
    Acked-by: Paul Mackerras <paulus@samba.org>
    Reviewed-by: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_hv_rmhandlers.S b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
index be2e8f6..398666e 100644
--- a/arch/powerpc/kvm/book3s_hv_rmhandlers.S
+++ b/arch/powerpc/kvm/book3s_hv_rmhandlers.S
@@ -313,6 +313,11 @@ kvm_no_guest:
  li r0, KVM_HWTHREAD_IN_NAP
  stb r0, HSTATE_HWTHREAD_STATE(r13)
 kvm_do_nap:
+ /* Clear the runlatch bit before napping */
+ mfspr r2, SPRN_CTRLF
+ clrrdi r2, r2, 1
+ mtspr SPRN_CTRLT, r2
+
  li r3, LPCR_PECE0
  mfspr r4, SPRN_LPCR
  rlwimi r4, r3, 0, LPCR_PECE0 | LPCR_PECE1
@@ -1953,8 +1958,13 @@ END_FTR_SECTION_IFCLR(CPU_FTR_ARCH_206)
 
  /*
   * Take a nap until a decrementer or external or doobell interrupt
-  * occurs, with PECE1, PECE0 and PECEDP set in LPCR
+  * occurs, with PECE1, PECE0 and PECEDP set in LPCR. Also clear the
+  * runlatch bit before napping.
   */
+ mfspr r2, SPRN_CTRLF
+ clrrdi r2, r2, 1
+ mtspr SPRN_CTRLT, r2
+
  li r0,1
  stb r0,HSTATE_HWTHREAD_REQ(r13)
  mfspr r5,SPRN_LPCR
-- 
1.7.1